lib/codecs/src/encoding/format/gelf.rs

```diff
@@ -0,0 +1,470 @@
+///   suggest. We've elected to take a more strict approach to maintain backwards compatability
+///   in the event that we need to change the behavior to be more relaxed, so that prior versions
+///   of vector will still work.
+///   The exception is that if 'Additional fields' are found to be missing an underscore prefix and
+///   are otherwise valid field names, we prepend the underscore.
+
+/// Errors that can occur during GELF serialization
+#[derive(Debug, Snafu)]
+pub enum GelfSerializerError {
+    #[snafu(display("LogEvent does not contain required field: {}", field))]
+    MissingField { field: String },
+    #[snafu(display(
+        "LogEvent contains a value with an invalid type. field = {} type = {} expected type = {}",
+        field,
+        actual_type,
+        expected_type
+    ))]
+    InvalidValueType {
+        field: String,
+        actual_type: String,
+        expected_type: String,
+    },
+    #[snafu(display("LogEvent contains an invalid field name. field = {}", field))]
+    InvalidFieldName { field: String },
+}
+
+/// Config used to build a `GelfSerializer`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct GelfSerializerConfig;
+
+impl GelfSerializerConfig {
+    /// Creates a new `GelfSerializerConfig`.
+    pub const fn new() -> Self {
+        Self
+    }
+
+    /// Build the `GelfSerializer` from this configuration.
+    pub fn build(&self) -> GelfSerializer {
+        GelfSerializer::new()
+    }
+
+    /// The data type of events that are accepted by `GelfSerializer`.
+    pub fn input_type() -> DataType {
+        DataType::Log
+    }
+
+    /// The schema required by the serializer.
+    pub fn schema_requirement() -> schema::Requirement {
+        // While technically we support `Value` variants that can't be losslessly serialized to
+        // JSON, we don't want to enforce that limitation to users yet.
+        schema::Requirement::empty()
+    }
+}
+
+/// Serializer that converts an `Event` to bytes using the GELF format.
+/// Spec: https://docs.graylog.org/docs/gelf
+#[derive(Debug, Clone)]
+pub struct GelfSerializer;
+
+impl GelfSerializer {
+    /// Creates a new `GelfSerializer`.
+    pub fn new() -> Self {
+        GelfSerializer
+    }
+
+    /// Encode event and represent it as JSON value.
+    pub fn to_json_value(&self, event: Event) -> Result<serde_json::Value, vector_core::Error> {
+        match event {
+            Event::Log(log) => {
+                if let Some(conformed) = self.to_gelf_event(&log)? {
+                    serde_json::to_value(&conformed)
+                } else {
+                    serde_json::to_value(&log)
+                }
+            }
+            Event::Metric(_) | Event::Trace(_) => {
+                panic!("GELF Serializer does not support Metric or Trace events.")
+            }
+        }
+        .map_err(|e| e.to_string().into())
+    }
+
+    /// Returns Error for invalid type
+    fn err_invalid_type(
```

Comment:
Not a big deal, but this function, and some of the below, don't seem to need `self`. Maybe they could be pulled out as private helper functions outside of this `impl` block?

Reply:
I'm surprised there isn't a clippy ding for these situations?

Reply:
Ah good catch , this happened to me previously as well . I was also surprised clippy didn't warn about it.

Reply:
Because `GelfSerializer` has no member variables, it seems most of those functions don't need to be in the `impl` block ... At one point I had a member variable there so it made sense. But even without it , I think I was just keeping them in the `impl` block for organization sake. I'll pull them out though.